src/Doctrine/BaseRelation.php

```diff
@@ -83,4 +86,34 @@ public function setMapInverseRelation(bool $mapInverseRelation)
 
         return $this;
     }
+
+    public function shouldAvoidSetter(): bool
+    {
+        return $this->avoidSetter;
+    }
+
+    public function avoidSetter(bool $avoidSetter = true)
+    {
+        $this->avoidSetter = $avoidSetter;
+
+        return $this;
+    }
+
+    public function getReturnType(): ?string
+    {
+        return $this->returnType;
+    }
+
+    public function setReturnType(string $returnType, $overriddenReturnType)
+    {
+        $this->returnType = $returnType;
+        $this->overriddenReturnType = $overriddenReturnType;
+
+        return $this;
+    }
+
+    public function isOverriddenReturnTypeNullable(): bool
+    {
+        return $this->overriddenReturnType;
+    }
```

Comment:
I think I may have confused you. Here is how I would expect the new parts of this class to look (I'm also changing the name "overridden" to "custom"):

```php
private $customReturnType;
private $isCustomReturnTypeNullable;

public function getCustomReturnType(): ?string
{
    return $this->customReturnType;
}

public function isCustomReturnTypeNullable(): bool
{
    return $this->isCustomReturnTypeNullable;
}

public function setCustomReturnType(string $customReturnType, bool $isNullable): void
{
    $this->customReturnType = $customReturnType;
    $this->isCustomReturnTypeNullable = $isNullable;
}
```

Then, in `ClassSourceManipulator`, when we use this, the code would be:

```php
        $this->addGetter(
            $relation->getPropertyName(),
            $typeHint,
            $relation->getCustomReturnType() ?: $typeHint,
            // getter methods always have nullable return values
            // unless this has been customized explicitly
            $relation->getCustomReturnType() ? $relation->isCustomReturnTypeNullable() : $typeHint,
            $relation->isOverriddenReturnTypeNullable()
        );
```

Reply:
For now, without modifying other parts of the MakerBundle, i think we should do things like this in `BaseRelation`
```php

    private $customReturnType;
    private $isCustomReturnTypeNullable = true;

    public function isCustomReturnTypeNullable(): bool
    {
        return $this->isCustomReturnTypeNullable;
    }


    public function setCustomReturnType(string $customReturnType, bool $isNullable)
    {
        $this->customReturnType = $customReturnType;
        $this->isCustomReturnTypeNullable = $isNullable;

        return $this;
    }

Reply:
And this in `ClassSourceManipulator`